gcc/rust/typecheck/rust-tyty-variance-analysis.cc

```diff
@@ -0,0 +1,540 @@
+}
+
+void
+solve ()
+{
+  auto ctx = GenericTyVisitorCtx::get ();
+  ctx.solve ();
+  ctx.debug_print_solutions ();
+}
+
+std::vector<Variance>
+query_generic_variance (const ADTType &type)
+{
+  auto ctx = GenericTyVisitorCtx::get ();
+  auto solution_index = ctx.lookup_type_index (type.get_orig_ref ());
+  rust_assert (solution_index.has_value ());
+  auto num_lifetimes = type.get_num_lifetime_params ();
+  auto num_types = type.get_num_type_params ();
+
+  std::vector<Variance> result;
+  for (size_t i = 0; i < num_lifetimes + num_types; ++i)
+    {
+      result.push_back (ctx.solutions[solution_index.value () + i]);
+    }
+
+  return result;
+}
+std::vector<Variance>
+query_type_variances (BaseType *type)
+{
+  TyVisitorCtx ctx;
+  return ctx.collect_variances (*type);
+}
+
+std::vector<Region>
+query_type_regions (BaseType *type)
+{
+  TyVisitorCtx ctx;
+  return ctx.collect_regions (*type);
+}
+
+Variance
+Variance::reverse () const
+{
+  switch (kind)
+    {
+    case BIVARIANT:
+      return bivariant ();
+    case COVARIANT:
+      return contravariant ();
+    case CONTRAVARIANT:
+      return covariant ();
+    case INVARIANT:
+      return invariant ();
+    }
+
+  rust_unreachable ();
+}
+
+Variance
+Variance::join (Variance lhs, Variance rhs)
+{
+  return {Kind (lhs.kind | rhs.kind)};
+}
+
+void
+Variance::join (Variance rhs)
+{
+  *this = join (*this, rhs);
+}
+
+Variance
+Variance::transform (Variance lhs, Variance rhs)
+{
+  switch (lhs.kind)
+    {
+    case BIVARIANT:
+      return bivariant ();
+    case COVARIANT:
+      return rhs;
+    case CONTRAVARIANT:
+      return rhs.reverse ();
+    case INVARIANT:
+      return invariant ();
+    }
+  rust_unreachable ();
+}
+
+std::string
+Variance::as_string () const
+{
+  switch (kind)
+    {
+    case BIVARIANT:
+      return "o";
+    case COVARIANT:
+      return "+";
+    case CONTRAVARIANT:
+      return "-";
+    case INVARIANT:
+      return "*";
+    }
+  rust_unreachable ();
+}
+
+GenericTyVisitorCtx *GenericTyVisitorCtx::instance = nullptr;
+
+GenericTyVisitorCtx &
+GenericTyVisitorCtx::get ()
+{
+  if (instance == nullptr)
+    instance = new GenericTyVisitorCtx ();
+
+  return *instance;
+}
+
+void
+GenericTyVisitorCtx::process_type (ADTType &type)
+{
+  rust_debug ("add_type_constraints: %s", type.as_string ().c_str ());
+
+  first_lifetime = lookup_or_add_type (type.get_orig_ref ());
+  first_type
+    = first_lifetime + type.get_used_arguments ().get_regions ().size ();
+
+  for (const auto &param : type.get_substs ())
+    param_names.push_back (
+      param.get_generic_param ().get_type_representation ().as_string ());
+
+  for (const auto &variant : type.get_variants ())
+    {
+      if (variant->get_variant_type () != VariantDef::NUM)
+	{
+	  for (const auto &field : variant->get_fields ())
+	    add_constraints_from_ty (field->get_field_type (), V::covariant ());
+	}
+    }
+
+  // Cleanup
+  first_lifetime = 0;
+  first_type = 0;
+  param_names.clear ();
```

Comment:
I think we could possibly avoid this with a non-global GenericTyVisitorCtx, but again it might cause more harm than good to make it non-global - so I'm trusting you